File: server/src/main/kotlin/org/bitlap/server/raft/MetaStateMachine.kt
package org.bitlap.server.raft

import com.alipay.sofa.jraft.Iterator
import com.alipay.sofa.jraft.Status
import com.alipay.sofa.jraft.core.StateMachineAdapter
import java.util.concurrent.atomic.AtomicLong

/**
 * Desc:
 *
 * Mail: chk19940609@gmail.com
 * Created by IceMimosa
 * Date: 2021/4/23
 */
class MetaStateMachine : StateMachineAdapter() {

    private val leaderTerm = AtomicLong(-1)

    override fun onApply(iter: Iterator) {
        // election only, do nothing
        while (iter.hasNext()) {
            println("On apply with term: ${iter.term} and index: ${iter.index}.")
            iter.next()
        }
    }

    override fun onLeaderStart(term: Long) {
        super.onLeaderStart(term)
        this.leaderTerm.set(term)
    }

    override fun onLeaderStop(status: Status?) {
        super.onLeaderStop(status)
        this.leaderTerm.set(-1L)
    }
}

File: server/src/main/scala/org/bitlap/server/raft/MetaStateMachine.scala
/* Copyright (c) 2022 bitlap.org */
package org.bitlap.server.raft

import com.alipay.sofa.jraft.core.StateMachineAdapter

import com.alipay.sofa.jraft
import com.alipay.sofa.jraft.Status

import java.util.concurrent.atomic.AtomicLong

/**
 * @author 梦境迷离
 * @version 1.0,2021/12/3
 */
class MetaStateMachine extends StateMachineAdapter {

  private val leaderTerm = new AtomicLong(-1)

  override def onApply(iterator: jraft.Iterator): Unit =
    while (iterator.hasNext)
      iterator.next()

  override def onLeaderStart(term: Long) {
    super.onLeaderStart(term)
    this.leaderTerm.set(term)
  }

  override def onLeaderStop(status: Status) {
    super.onLeaderStop(status)
    this.leaderTerm.set(-1L)
  }
}
